### PR TITLE
feat: LLM cost visibility per feature (#270)

### PR DESCRIPTION
**Pull request by markus-lassfolk**
## Summary

- **`services/model-pricing.ts`**: Static pricing table for 14 models across OpenAI, Google, Anthropic, MiniMax with `estimateCost()` helper
- **`backends/cost-tracker.ts`**: `CostTracker` class backed by `llm_cost_log` table in the existing `memory.db` SQLite file (shares `FactsDB` connection, zero extra DB files)
- **`services/chat.ts`**: Optional `feature?` and `costTracker?` params on `chatComplete` and `chatCompleteWithRetry` — records `resp.usage` transparently, fully backward compatible
- **Config**: `CostTrackingConfig` type + `costTracking` field in `HybridMemoryConfig` (enabled by default, `retainDays=90`, `pruneInNightlyCycle=true`)
- **CLI `cost-report`**: `openclaw hybrid-mem cost-report [--days N] [--model] [--feature <name>] [--csv]` — tabular breakdown by feature or model
- **`verify` output**: New "Cost Tracking" section showing 7-day spend + top spenders; "Estimated Monthly Cost by Mode" section (essential/normal/expert/full)
- **Nightly prune**: Integrated in dream-cycle to prune old entries per `retainDays` config
- **29 new tests** across `tests/model-pricing.test.ts` (12) and `tests/cost-tracker.test.ts` (17)

## Architecture

```
LLM call
  └─ chatComplete({ feature: 'auto-classify', costTracker })
       └─ resp.usage → CostTracker.record()
            └─ INSERT INTO llm_cost_log (feature, model, input_tokens, ...)
                 └─ estimated_cost_usd = estimateCost(model, in, out)

openclaw hybrid-mem cost-report --days 7
  └─ CostTracker.getReport() → tabular output by feature

openclaw hybrid-mem verify
  └─ cost tracking section: event count + last 7d spend + top spenders
  └─ estimated monthly cost by mode
```

## Test plan

- [x] `npm test` — all 111 test files pass (2699 tests, +29 new)
- [x] `CostTracker.record()` inserts row with correct cost calculation
- [x] `CostTracker.getReport()` aggregates by feature, filters by window
- [x] `CostTracker.pruneOldEntries()` deletes old entries, returns count
- [x] `estimateCost()` correct for known models, null for unknown
- [x] Schema idempotent — `CostTracker` constructor safe to call twice
- [x] Backward compatible — `chatComplete` without `feature`/`costTracker` works unchanged

⚠️ Prices are estimates based on published model pricing. Actual costs may vary.
Embedding calls (text-embedding-3-small) are not included.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Instruments all `chat.completions.create` calls via the OpenAI proxy to asynchronously log usage into SQLite, which could impact LLM-call stability/performance if the heuristic labeling or DB writes misbehave (though failures are swallowed). Adds new CLI/config surface area and nightly pruning that could delete data unexpectedly if misconfigured.
> 
> **Overview**
> Adds **LLM cost visibility per feature** by introducing a `CostTracker` backed by new `llm_cost_log`/`llm_savings_log` tables in the existing SQLite `memory.db`, plus a static `MODEL_PRICING` table to estimate USD costs.
> 
> Instruments the multi-provider OpenAI proxy to **fire-and-forget log token usage, duration, success/failure, and inferred feature labels** (AsyncLocalStorage label or prompt-content heuristics), and wires this into config (`costTracking`), `verify` output, and a new `openclaw hybrid-mem cost-report` command (feature/model breakdown, CSV, and mode cost estimates).
> 
> Adds automation “savings” logging for several lifecycle tasks, and prunes old cost/savings entries during the nightly dream-cycle based on retention settings; includes comprehensive unit tests for pricing, tracking, aggregation, and pruning.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a590699b290f52548d59420629e8681e58a4fc49. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: extensions/memory-hybrid/services/chat.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/chat.ts]: ### 🔴 CRITICAL: Zero callers are instrumented

**Critical**

The infrastructure for cost tracking is built, but none of the actual callers (`auto-classifier.ts`, `query-expander.ts`, `reflection.ts`, etc.) have been updated to pass the `feature` and `costTracker` parameters. Furthermore, direct `openai.chat.completions.create` callers are completely uninstrumented. Until the callers are updated, zero data will be recorded.

File: extensions/memory-hybrid/services/chat.ts [hunk 2]
```diff
@@ -93,6 +98,19 @@ export async function chatComplete(opts: {
...
+          feature: opts.feature,
+          model,
+          inputTokens: resp.usage.prompt_tokens ?? 0,
+          outputTokens: resp.usage.completion_tokens ?? 0,
+          success: true,
+        });
+      } catch {
+        // Cost tracking is best-effort; never let it break the caller
+      }
+    }
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/chat.ts]: ### 🟡 WARNING: Failed calls are not tracked

**Warning**

The design spec specifies `success: 0 if call failed (still costs tokens)`. However, if `openai.chat.completions.create` throws an error, execution jumps to the catch block and the cost tracking is completely bypassed. Consider catching the error, attempting to extract any partial usage data if available, and logging the failure before rethrowing.

File: extensions/memory-hybrid/cli/handlers.ts [hunk 3]
```diff
@@ -4384,3 +4418,138 @@ export async function runToolEffectivenessForCli(
...
+    log([
+      "Total".padEnd(colW[0]!),
+      String(total.calls).padStart(colW[1]!),
+      fmtNum(total.inputTokens).padStart(colW[2]!),
+      fmtNum(total.outputTokens).padStart(colW[3]!),
+      fmtCost(total.estimatedCostUsd).padStart(colW[4]!),
+    ].join("  "));
+  } else {
+    // Feature breakdown
+    const report = costTracker.getReport({ days, feature: opts.feature });
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/cli/handlers.ts]: ### 🟢 SUGGESTION: Dynamic column width

**Suggestion**

The `colW` array uses fixed widths (e.g., 30 for Feature). If a new feature label exceeds 30 characters in the future, the table columns will misalign. Consider calculating the max feature length dynamically.

**Review by markus-lassfolk — COMMENTED** [review]
## 🤖 GPT UX & Testing Review — PR #272 (Cost Visibility)

### Summary
- 🔴 **Critical:** No LLM callers are instrumented with {feature, costTracker} yet → **zero cost data will ever be recorded**. UX-wise: user runs cost-report/verify and always sees *no data yet* forever.
- 🟡 **Warning:** Unknown-model pricing is silently treated as **$0** (SQL COALESCE(SUM(estimated_cost_usd), 0)) → undercounted totals with no visibility.
- 🟡 **Warning:** CLI output is close, but a few edge cases remain (column truncation, inconsistent precision, CSV escaping, and messaging when tracker is present but no instrumentation exists).
- 🟢 **Suggestion:** Add minimal end-to-end tests proving one real call path records a row + verify/cost-report shows it.

### What users experience today
- Cost tracker initializes and schema exists.
- hybrid-mem cost-report almost always prints “No LLM cost data…”.
- verify shows Cost Tracking: ✅ on (no data yet …) even if the system has been running for weeks.

### Practical phased rollout (so this lands safely)
1) Instrument 3–5 highest-frequency callers immediately (auto-classify, query-expansion/HyDE, reranking, reflection, distill) and ship.
2) Add a single wrapper (or OpenAI client adapter) so direct openai.chat.completions.create() callers can’t bypass cost tracking.
3) Add reporting for unpriced/unknown models and optionally “untracked LLM calls detected”.

Signed: 🤖 GPT Council Member

File: extensions/memory-hybrid/services/model-pricing.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/**
+ * Static LLM model pricing table for cost estimation.
+ *
+ * ⚠️ DISCLAIMER: Prices are rough estimates based on publicly published
+ * rates as of early 2025. Actual billing may differ due to promotions,
+ * batch discounts, caching, or provider changes. Do not use for billing.
+ */
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/model-pricing.ts]: ### 🟢 Pricing disclaimer says "early 2025" — should be 2026

**Suggestion**

Minor: the disclaimer reads "as of early 2025" but the PR is from March 2026. Update to "as of early 2026" or make it date-agnostic: "based on publicly listed rates at time of writing".

File: extensions/memory-hybrid/services/model-pricing.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * Look up pricing for a model. Returns null if the model is not in the table.
+ * Model name is matched case-insensitively.
+ */
+export function getModelPricing(model: string): ModelPricing | null {
+  // Direct match first
+  if (model in MODEL_PRICING) return MODEL_PRICING[model]!;
+  // Case-insensitive match
+  const lower = model.toLowerCase();
+  for (const [key, pricing] of Object.entries(MODEL_PRICING)) {
+    if (key.toLowerCase() === lower) return pricing;
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/model-pricing.ts]: ### 🟢 Case-insensitive lookup iterates entire table on every miss

**Suggestion**

The fallback loop through `Object.entries(MODEL_PRICING)` on case mismatch runs on every call with a model not in the table (which will be frequent for unknown models). Consider building a lowercased lookup map once at module load:

```typescript
const LOWER_PRICING = new Map(
  Object.entries(MODEL_PRICING).map(([k, v]) => [k.toLowerCase(), v])
);
```

Not a perf crisis (14 entries), but good hygiene for a function called per-LLM-call.

File: extensions/memory-hybrid/tests/cost-tracker.test.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
+import { describe, it, expect, beforeEach, afterEach } from "vitest";
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/tests/cost-tracker.test.ts]: ### 🟢 No tests for `runCostReportForCli` or verify output section

**Suggestion**

The CostTracker and model-pricing tests are solid (29 tests total). Missing coverage:

1. **`runCostReportForCli`** — no tests for the CLI handler that formats the report. Should test: empty data, feature filter, `--model` flag, `--csv` output, the model summary line
2. **Verify section** — no tests for the cost tracking section added to `runVerifyForCli`. Should test: costTracker=null, zero data, with data, disabled-by-config
3. **`chatComplete` integration** — no test verifying that `chatComplete` actually calls `costTracker.record()` when both `feature` and `costTracker` are provided (the try/catch branch is untested)

These are the user-facing surfaces — if formatting breaks, users see garbled output.

File: extensions/memory-hybrid/services/model-pricing.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  inputPer1M: number;
+  /** USD per 1M output tokens */
+  outputPer1M: number;
+}
+
+/** Static pricing table. Keys match the provider/model format used in LLM config. */
+export const MODEL_PRICING: Record<string, ModelPricing> = {
+  // OpenAI
+  "openai/gpt-4.1-nano": { inputPer1M: 0.10, outputPer1M: 0.40 },
+  "openai/gpt-4.1-mini": { inputPer1M: 0.40, outputPer1M: 1.60 },
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/model-pricing.ts]: ### 🟢 Pricing table missing some active models

**Suggestion**

Good coverage of 14 models across 4 providers. A few potentially missing:
- `openai/gpt-4.1` (the full-size 4.1, not just nano/mini)
- `google/gemini-2.5-flash` (newer Gemini models)
- `anthropic/claude-sonnet-4-5` / `claude-haiku-4` (if in use)

The null-cost fallback handles unknown models gracefully, and the `unknownModels` tracking in `getReport()` will surface gaps at runtime. Nice-to-have, not blocking.

File: extensions/memory-hybrid/cli/handlers.ts (hunk 3 above)
Comment (markus-lassfolk) [reply extensions/memory-hybrid/cli/handlers.ts]: Addressed in commit 7ddd36b — column widths were already dynamic (`Math.max(20, ...items.map(r => r.feature.length)) + 2`); the report was improved with a percentage column, emoji summary headers, and a `--format compact` option for terse output.

File: extensions/memory-hybrid/services/chat.ts [hunk 8]
```diff
@@ -67,6 +68,10 @@ export async function chatComplete(opts: {
   timeoutMs?: number;
   /** When aborted (e.g. parent timeout), the request is cancelled and no retry is needed. */
   signal?: AbortSignal;
+  /** Feature label for cost tracking (e.g. 'auto-classify', 'query-expansion'). */
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/chat.ts]: Fixed in commits d954cba + 4273b51 + 7ddd36b — all cost tracking is now proxy-based (no call sites need to pass anything). Additionally, `chatComplete` now accepts an optional `feature` param that wraps the call in `withCostFeature()` for precise attribution, and the unused `costTracker` param was removed from both `chatComplete` and `chatCompleteWithRetry`.

File: extensions/memory-hybrid/setup/init-databases.ts [hunk 9]
```diff
@@ -385,6 +387,11 @@ export function initializeDatabases(
     api.logger.info("memory-hybrid: verification store enabled");
   }
 
+  // Initialize CostTracker — always enabled; recording gated by cfg.costTracking.enabled (Issue #270)
+  // Shares FactsDB's SQLite connection (same memory.db file, avoids a second DB handle).
+  const costTracker = new CostTracker(factsDb.getRawDb());
+  api.logger.info("memory-hybrid: LLM cost tracker initialized");
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/setup/init-databases.ts]: Implemented in commit d954cba — `buildMultiProviderOpenAI()` in `setup/init-databases.ts` is the single instrumentation point. The proxy intercepts `chat.completions.create`, infers feature labels (AsyncLocalStorage first, heuristics as fallback), and records cost fire-and-forget without blocking the LLM call.

File: extensions/memory-hybrid/setup/init-databases.ts [hunk 10]
```diff
@@ -385,6 +387,11 @@ export function initializeDatabases(
     api.logger.info("memory-hybrid: verification store enabled");
   }
 
+  // Initialize CostTracker — always enabled; recording gated by cfg.costTracking.enabled (Issue #270)
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/setup/init-databases.ts]: Fixed in commit d954cba — both the proxy init and the verify section now gate on `cfg.costTracking?.enabled !== false`. The `CostTracker` is only created when enabled, and verify correctly shows ⏸ Disabled when the flag is false.

File: extensions/memory-hybrid/cli/handlers.ts [hunk 11]
```diff
@@ -976,6 +979,37 @@ export async function runVerifyForCli(
     }
   }
 
+  // ───── Cost Tracking ─────
+  log("\n───── Cost Tracking ─────");
+  if (ctx.costTracker && ctx.cfg.costTracking.enabled !== false) {
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/cli/handlers.ts]: Fixed in commit d954cba — verify checks `ctx.cfg.costTracking.enabled` first. If false, it shows OFF. If true but `costTracker` is null (should not happen in normal flow), it falls back to OFF. The ✅/⏸ output was further improved in 7ddd36b.

File: extensions/memory-hybrid/setup/init-databases.ts [hunk 12]
```diff
@@ -18,6 +18,7 @@ import { setMemoryCategories, getMemoryCategories } from "../config.js";
 import { migrateCredentialsToVault, CREDENTIAL_REDACTION_MIGRATION_FLAG } from "../services/credential-migration.js";
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/setup/init-databases.ts]: Thank you for verifying — confirmed in all commits.

File: extensions/memory-hybrid/cli/handlers.ts [hunk 13]
```diff
@@ -84,6 +84,7 @@ import {
 } from "../utils/constants.js";
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/cli/handlers.ts]: Thank you for verifying — `cost-report` was further improved in 7ddd36b with emoji headers, summary line, percentage column, and `--format compact` option.
